### PR TITLE
ARIA-284 Cleanup and optimize the task execution

### PR DESCRIPTION
**Pull request by mxmrlv**

### CODE REVIEW

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

File: aria/orchestrator/workflow_runner.py [hunk 1]
```diff
@@ -96,8 +96,9 @@ def __init__(self, model_storage, resource_storage, plugin_manager,
 
         if not self._is_resume:
             workflow_fn = self._get_workflow_fn()
-            tasks_graph = workflow_fn(ctx=self._workflow_context, **execution_inputs_dict)
-            compile.create_execution_tasks(self._workflow_context, tasks_graph, executor.__class__)
+            self._tasks_graph = workflow_fn(ctx=self._workflow_context, **execution_inputs_dict)
+            compile.GraphCompiler(self._workflow_context, executor.__class__).compile(
```
Comment (mxmrlv) [inline aria/orchestrator/workflow_runner.py]: graph_compiler

File: aria/orchestrator/workflows/core/compile.py [hunk 2]
```diff
@@ -18,99 +18,105 @@
...
-    tasks = []
-    for dependency in dependencies:
-        if getattr(dependency, 'actor', False):
-            # This is
-            dependency_name = dependency.id
-        else:
-            dependency_name = _end_graph_suffix(dependency.id)
-        tasks.extend(task for task in execution.tasks if task._api_id == dependency_name)
-    return tasks or default
+# TODO: is class really needed?
```
Comment (mxmrlv) [inline aria/orchestrator/workflows/core/compile.py]: remove

File: aria/orchestrator/workflows/core/engine.py [hunk 3]
```diff
@@ -45,22 +45,23 @@ def execute(self, ctx, resuming=False):
...
                 if cancel:
                     break
-                for task in self._ended_tasks(ctx, executing_tasks):
-                    self._handle_ended_tasks(ctx, task, executing_tasks)
-                for task in self._executable_tasks(ctx):
-                    self._handle_executable_task(ctx, task, executing_tasks)
-                if self._all_tasks_consumed(ctx):
+                for task in task_tracker.ended_tasks:
+                    self._handle_ended_tasks(task)
+                    task_tracker.finished_(task)
```
Comment (mxmrlv) [inline aria/orchestrator/workflows/core/engine.py]: remove _

File: aria/orchestrator/workflows/core/engine.py [hunk 4]
```diff
@@ -129,16 +103,70 @@ def _handle_executable_task(self, ctx, task, executing_tasks):
...
+    @property
+    def ended_tasks(self):
+        for task in self.executing_tasks:
+            if task.has_ended():
+                yield task
+
+    @property
+    def executable_tasks(self):
+        now = datetime.utcnow()
+        # we need both list since retrying task are in the executing task list.
```
Comment (mxmrlv) [inline aria/orchestrator/workflows/core/engine.py]: lists

**Pull request merged**